File: scanner/tests/fixtures/vulnerable_app.go
package main

import (
	"database/sql"
	"fmt"
	"net/http"
	"os/exec"
)

var db *sql.DB

// SQL Injection
func getUser(w http.ResponseWriter, r *http.Request) {
	id := r.URL.Query().Get("id")
	rows, _ := db.Query("SELECT * FROM users WHERE id = " + id)
	defer rows.Close()
	fmt.Fprintf(w, "result: %v", rows)
}

// Command Injection
func ping(w http.ResponseWriter, r *http.Request) {
	host := r.URL.Query().Get("host")
	cmd := exec.Command("ping", "-c", "1", host)
	output, _ := cmd.Output()
	fmt.Fprintf(w, "%s", output)
}

// XSS via direct write
func greet(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("name")
	fmt.Fprintf(w, "<h1>Hello %s</h1>", name)
}

func validateInput(data string) string {
	if data == "" {
		panic("empty input")
	}
	return data
}

func queryDatabase(id string) *sql.Rows {
	rows, _ := db.Query("SELECT * FROM users WHERE id = " + id)
	return rows
}

func getUser2(w http.ResponseWriter, r *http.Request) {
	id := validateInput(r.URL.Query().Get("id"))
	rows := queryDatabase(id)
	defer rows.Close()
}

func main() {
	http.HandleFunc("/users", getUser)
	http.HandleFunc("/ping", ping)
	http.HandleFunc("/greet", greet)
	http.HandleFunc("/users2", getUser2)
	http.ListenAndServe(":8080", nil)
}

File: scanner/tests/fixtures/vulnerable_app.js
const express = require('express');
const mysql = require('mysql');
const { exec } = require('child_process');
const fs = require('fs');

const app = express();
const db = mysql.createConnection({ host: 'localhost', user: 'root', database: 'app' });

// SQL Injection
app.get('/users', (req, res) => {
    const id = req.query.id;
    db.query("SELECT * FROM users WHERE id = " + id, (err, rows) => {
        res.json(rows);
    });
});

// Command Injection
app.get('/ping', (req, res) => {
    const host = req.query.host;
    exec("ping -c 1 " + host, (err, stdout) => {
        res.send(stdout);
    });
});

// XSS via innerHTML (pattern-based)
app.get('/greet', (req, res) => {
    const name = req.query.name;
    res.send(`<div id="output"></div><script>document.getElementById('output').innerHTML = '${name}'</script>`);
});

// Path Traversal
app.get('/file', (req, res) => {
    const filename = req.query.file;
    fs.readFile('/data/' + filename, 'utf-8', (err, data) => {
        res.send(data);
    });
});

function validateInput(data) {
    if (!data) throw new Error('empty');
    return data.trim();
}

function queryUser(id) {
    return db.query("SELECT * FROM users WHERE id = " + id);
}

app.get('/users2', (req, res) => {
    const id = validateInput(req.query.id);
    const result = queryUser(id);
    res.json(result);
});

app.listen(3000);
